### PR TITLE
Add support of build mounts

### PR DESCRIPTION
**Pull request by stanislav-zaprudskiy**
* https://docs.docker.com/build/building/secrets/
* https://docs.podman.io/en/latest/markdown/podman-run.1.html#secret-secret-opt-opt

A possible implementation of https://github.com/ansible/ansible-builder/issues/196 (but is not related to https://github.com/ansible/ansible-builder/issues/195)

### CODE REVIEW

**Comment by stanislav-zaprudskiy** [comment]
I thought of extending the user definition file schema to allow passing mounts and secrets too. Thus, these lengthy arguments would persist in file, instead of needing to re-type them every time.

However, this doesn't really come along with _somewhat_ similar `build_arg_defaults` - which only allows passing the `ansible-builder` provided arguments, and not any random ones.

**Comment by stanislav-zaprudskiy** [comment]
Another thing is that perhaps updating the integration tests and docs would be needed.

**Comment by stanislav-zaprudskiy** [comment]
I force-pushed to incorporate latest `ansible:devel` changes causing merge conflicts previously.

**Comment by Kaezon** [comment]
I just did some testing, and I ran into a problem where ansible-builder was selecting the podman runtime while I was specifying `--ssh`. Since podman doesn't support secret mounts, there should probably be a check to bail out if the selected runtime is podman.

Bit of a side-note, I noticed that the changed files have since moved. Some code was moved into `containerfile.py` as well.
It only required a few minutes of copy-pasting to correct though.

**Comment by stanislav-zaprudskiy** [comment]
I plan to update the PR soon, just haven't found time yet to catch up with the recent `ansible-builder` changes.

**Comment by stanislav-zaprudskiy** [comment]
@Shrews, @Akasurde, would you give this a review?

**Comment by Shrews** [comment]
I'll take a look at this shortly. Thanks for keeping it updated.

File: src/ansible_builder/containerfile.py [hunk 1]
```diff
@@ -207,6 +209,7 @@ def write(self) -> None:
         """
         with open(self.path, 'w') as f:
             for step in self.steps:
+                step = self._add_mounts_for_run_step(step)
```
Comment (Shrews) [inline src/ansible_builder/containerfile.py:212]: This is adding the list of mounts to _every_ `RUN` command. Definitely do not like that as it's very much overkill. I think if we are going to allow secrets, then we should assume that the user is going to manually add `RUN` statements that will use those secrets to the EE definition file under `additional_run_steps` where they will manually specify the mounts, which means we could eliminate the `--mounts` option entirely.

**Review by Shrews — CHANGES_REQUESTED** [review]
So, we definitely need some documentation, ideally with examples, about these new features/options.

How the mounts is being handled isn't very clean, IMO. Not every RUN command needs those mounts. I can only see manually inserted RUN commands (via `additional_build_steps`) ever needing to access the secrets. Unless I've missed something?

I'm not sure about the `--ssh` option, but for the `--secret` option, I believe we could create an integration test for that. It could add a secret from a file, cat that secret from the RUN command we add to the EE, and we could check the build output for the secret contents. This would validate that it works all the way through, unlike the unit tests.

**Comment by Shrews** [comment]
Come to think of it, I think we could wrap both of these new options up in a more generic --extra-cli-opts argument (or similar) that we just pass through to the build command.

**Comment by Shrews** [comment]
I was able to quickly hack up a proof of concept for my suggestion, with a working test proving that secrets work in PR #677. I may need to clean that up a bit with some comments, error checking, unit tests, adding docs, etc (just quickly vomited some code to see if it would work).

**Comment by Shrews** [comment]
Now that PR #677 is merged and released in version `3.1`, I'm going to go ahead and close this. I added a [scenario doc](https://ansible.readthedocs.io/projects/builder/en/3.1.0/scenario_guides/scenario_secret_passing/) that shows how to use this new feature to pass a secret.

**Pull request closed without merging**